File: ios/GuardApp/Core/Identity/LocalIdentityStore.swift
import Foundation
import Security

struct LocalIdentity: Equatable {
    let deviceId: String
    let groupSecret: Data?
}

enum LocalIdentityError: Error, Equatable {
    case invalidStoredDeviceId
    case keychainUnhandledStatus(OSStatus)
}

protocol SecureIdentityStorage {
    func read(key: String) throws -> Data?
    func write(_ data: Data, key: String) throws
    func delete(key: String) throws
}

final class LocalIdentityStore {
    private enum Keys {
        static let deviceId = "deviceId"
        static let groupSecret = "groupSecret"
    }

    private let storage: SecureIdentityStorage
    private let idFactory: () -> String

    init(
        storage: SecureIdentityStorage,
        idFactory: @escaping () -> String = { UUID().uuidString }
    ) {
        self.storage = storage
        self.idFactory = idFactory
    }

    func loadOrCreateIdentity() throws -> LocalIdentity {
        let deviceId = try loadOrCreateDeviceId()
        let groupSecret = try storage.read(key: Keys.groupSecret)
        return LocalIdentity(deviceId: deviceId, groupSecret: groupSecret)
    }

    func saveGroupSecret(_ secret: Data) throws {
        try storage.write(secret, key: Keys.groupSecret)
    }

    func clear() throws {
        try storage.delete(key: Keys.deviceId)
        try storage.delete(key: Keys.groupSecret)
    }

    private func loadOrCreateDeviceId() throws -> String {
        if let stored = try storage.read(key: Keys.deviceId) {
            guard let value = String(data: stored, encoding: .utf8), !value.isEmpty else {
                throw LocalIdentityError.invalidStoredDeviceId
            }
            return value
        }

        let created = idFactory()
        try storage.write(Data(created.utf8), key: Keys.deviceId)
        return created
    }
}

final class KeychainIdentityStorage: SecureIdentityStorage {
    private let service: String

    init(service: String = "com.deviceguard.identity") {
        self.service = service
    }

    func read(key: String) throws -> Data? {
        var query = baseQuery(key: key)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne

        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)

        if status == errSecItemNotFound {
            return nil
        }
        guard status == errSecSuccess else {
            throw LocalIdentityError.keychainUnhandledStatus(status)
        }
        return item as? Data
    }

    func write(_ data: Data, key: String) throws {
        var query = baseQuery(key: key)
        let attributes = [kSecValueData as String: data]

        let updateStatus = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
        if updateStatus == errSecSuccess {
            return
        }
        guard updateStatus == errSecItemNotFound else {
            throw LocalIdentityError.keychainUnhandledStatus(updateStatus)
        }

        query[kSecValueData as String] = data
        query[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        let addStatus = SecItemAdd(query as CFDictionary, nil)
        guard addStatus == errSecSuccess else {
            throw LocalIdentityError.keychainUnhandledStatus(addStatus)
        }
    }

    func delete(key: String) throws {
        let status = SecItemDelete(baseQuery(key: key) as CFDictionary)
        guard status == errSecSuccess || status == errSecItemNotFound else {
            throw LocalIdentityError.keychainUnhandledStatus(status)
        }
    }

    private func baseQuery(key: String) -> [String: Any] {
        [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key
        ]
    }
}

final class InMemoryIdentityStorage: SecureIdentityStorage {
    private var values: [String: Data] = [:]

    func read(key: String) throws -> Data? {
        values[key]
    }

    func write(_ data: Data, key: String) throws {
        values[key] = data
    }

    func delete(key: String) throws {
        values.removeValue(forKey: key)
    }
}

File: android/app/src/main/java/com/deviceguard/core/identity/LocalIdentityStore.kt
package com.deviceguard.core.identity

import java.security.SecureRandom
import java.util.UUID

data class LocalIdentity(
    val deviceId: String,
    val groupSecret: ByteArray?
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is LocalIdentity) return false
        return deviceId == other.deviceId && groupSecret.contentEquals(other.groupSecret)
    }

    override fun hashCode(): Int {
        var result = deviceId.hashCode()
        result = 31 * result + (groupSecret?.contentHashCode() ?: 0)
        return result
    }
}

interface SecureIdentityStorage {
    fun read(key: String): ByteArray?
    fun write(key: String, value: ByteArray)
    fun delete(key: String)
}

class LocalIdentityStore(
    private val storage: SecureIdentityStorage,
    private val idFactory: () -> String = { UUID.randomUUID().toString() },
    private val secretFactory: (Int) -> ByteArray = { size ->
        ByteArray(size).also { SecureRandom().nextBytes(it) }
    }
) {
    fun loadOrCreateIdentity(): LocalIdentity {
        val deviceId = loadOrCreateDeviceId()
        return LocalIdentity(
            deviceId = deviceId,
            groupSecret = storage.read(KEY_GROUP_SECRET)
        )
    }

    fun saveGroupSecret(secret: ByteArray) {
        require(secret.isNotEmpty()) { "groupSecret must not be empty" }
        storage.write(KEY_GROUP_SECRET, secret.copyOf())
    }

    fun createGroupSecret(size: Int = DEFAULT_GROUP_SECRET_SIZE): ByteArray {
        require(size >= MIN_GROUP_SECRET_SIZE) { "groupSecret must be at least $MIN_GROUP_SECRET_SIZE bytes" }
        return secretFactory(size)
    }

    fun clear() {
        storage.delete(KEY_DEVICE_ID)
        storage.delete(KEY_GROUP_SECRET)
    }

    private fun loadOrCreateDeviceId(): String {
        val stored = storage.read(KEY_DEVICE_ID)
            ?.toString(Charsets.UTF_8)
            ?.takeIf { it.isNotBlank() }
        if (stored != null) return stored

        val created = idFactory()
        storage.write(KEY_DEVICE_ID, created.toByteArray(Charsets.UTF_8))
        return created
    }

    companion object {
        private const val KEY_DEVICE_ID = "deviceId"
        private const val KEY_GROUP_SECRET = "groupSecret"
        private const val DEFAULT_GROUP_SECRET_SIZE = 32
        private const val MIN_GROUP_SECRET_SIZE = 16
    }
}

class InMemoryIdentityStorage : SecureIdentityStorage {
    private val values = mutableMapOf<String, ByteArray>()

    override fun read(key: String): ByteArray? = values[key]?.copyOf()

    override fun write(key: String, value: ByteArray) {
        values[key] = value.copyOf()
    }

    override fun delete(key: String) {
        values.remove(key)
    }
}
